File: python/programmers/dfs-bfs/타겟넘버.py
'''
* 🤷‍♂️ Created by wwlee94 on 2020.02.07
https://programmers.co.kr/learn/courses/30/lessons/43165

- 문제 풀이 접근 - 
'+', '-' 연산 두가지 방법을 선택해 리스트의 합이 target과 같은 값을 출력하는 경우를 카운트 하는 문제
따라서 재귀 함수로 '+' 연산 한번 '-' 연산 한번 진행하면서 
마지막 원소에 접근할 때 모든 원소의 값이 target과 동일한지 검사  
같으면 answer += 1로 카운팅해주면 끝 !
'''
def solution(numbers, target):
    answer = 0
    len_numbers = len(numbers)

    def DFS(idx=0):
        if idx < len_numbers:
            numbers[idx] *= 1
            DFS(idx+1)

            numbers[idx] *= -1
            DFS(idx+1)

        elif sum(numbers) == target:
            nonlocal answer
            answer += 1
            
    DFS()
    return answer

# 2020.07.22 - 최근에 풀었지만, 조금 더 느린 풀이
def solution(numbers, target):
    answer = 0
    candidates = []
    
    def DFS(number=0, i=0):
        nonlocal answer

        if i != 0:
            candidates.append(number)
        
        if len(candidates) == len(numbers):
            if sum(candidates) == target:
                answer += 1
            return
    
        DFS(numbers[i], i+1)
        candidates.pop()
        DFS(numbers[i] * -1, i+1)
        candidates.pop()
        
    DFS()
    return answer

File: java/programmers/dfs-bfs/타겟넘버.java
/*
* 🤷‍♂️ Created by wwlee94 on 2020.02.15
https://programmers.co.kr/learn/courses/30/lessons/43165

- Tip -
1. class 부분에 전역으로 사용할 static 변수 or 일반 변수 만들어서 사용해도 됨
EX) 
class Solution{
    static int[] numbers;
    // int[] numbers;
    public int solution(int[] numbers, ...){
        this.numbers = numbers //this -> Solution 클래스를 가리킬 듯
    }

2. 향상된 for문 사용
EX)
ArrayList<Integer> temp = new ArrayList<>();
...
for(int value : temp) System.out.print(value+" "); //ArrayList 출력
*/
import java.util.*;
class Solution {
    int answer = 0;
    ArrayList<Integer> temp = new ArrayList<>();
    public void searchDFS(int[] numbers, int floor, int target){
        if (floor < numbers.length){
            int val = numbers[floor];
        
            temp.add(val);
            searchDFS(numbers, floor + 1, target);
            temp.remove(temp.size()-1);

            temp.add(-val);
            searchDFS(numbers, floor + 1, target);
            temp.remove(temp.size()-1);
        }
        else if (floor == numbers.length){
            long sum = temp.stream()
                .mapToLong(Integer::longValue)
                .sum();
            if (sum == target) answer += 1;
        }
    }
    public int solution(int[] numbers, int target) {
        searchDFS(numbers, 0, target);        
        return answer;
    }
}